src/uu/cat/src/cat.rs

```diff
@@ -297,125 +377,82 @@ fn get_input_type(path: &str) -> CatResult<InputType> {
-fn write_fast(files: Vec<String>) -> CatResult<()> {
-    let mut writer = stdout();
-    let mut in_buf = [0; 1024 * 64];
-    let mut error_count = 0;
-
-    for file in files {
-        match open(&file[..]) {
-            Ok(mut handle) => {
-                while let Ok(n) = handle.reader.read(&mut in_buf) {
-                    if n == 0 {
-                        break;
-                    }
-                    writer.write_all(&in_buf[..n]).context(&file[..])?;
-                }
-            }
-            Err(error) => {
-                writeln!(&mut stderr(), "{}", error)?;
-                error_count += 1;
-            }
+    // If we're not on Linux or Android, or the splice() call failed,
+    // fall back on slower writing.
+    let mut buf = [0; 1024 * 64];
+    while let Ok(n) = handle.reader.read(&mut buf) {
+        if n == 0 {
+            break;
         }
-    }
-
-    match error_count {
-        0 => Ok(()),
-        _ => Err(CatError::EncounteredErrors(error_count)),
-    }
+        stdout_lock.write_all(&buf[..n])?;
+    };
+    Ok(())
 }
 
-/// State that persists between output of each file
-struct OutputState {
-    /// The current line number
-    line_number: usize,
+/// This function is called from `write_fast()` on Linux and Android. The
+/// function `splice()` is used to move data between two file descriptors
+/// without copying between kernel- and userspace. This results in a large
+/// speedup.
+#[cfg(any(target_os = "linux", target_os = "android"))]
+#[inline]
+fn write_fast_using_splice<R: Read>(
+    handle: &mut InputHandle<R>,
+    writer: RawFd,
+) -> Result<(), nix::Error> {
+    const BUF_SIZE: usize = 1024 * 16;
 
-    /// Whether the output cursor is at the beginning of a new line
-    at_line_start: bool,
-}
+    let (pipe_rd, pipe_wr) = pipe()?;
 
-/// Writes files to stdout with `options` as configuration.  Returns
-/// `Ok(())` if no errors were encountered, or an error with the
-/// number of errors encountered.
-///
-/// # Arguments
-///
-/// * `files` - There is no short circuit when encountering an error
-/// reading a file in this vector
-fn write_lines(files: Vec<String>, options: &OutputOptions) -> CatResult<()> {
-    let mut error_count = 0;
-    let mut state = OutputState {
-        line_number: 1,
-        at_line_start: true,
-    };
-
-    for file in files {
-        if let Err(error) = write_file_lines(&file, options, &mut state) {
-            writeln!(&mut stderr(), "{}", error).context(&file[..])?;
-            error_count += 1;
+    loop {
+        let res = splice(
+            handle.file_descriptor,
+            None,
+            pipe_wr,
+            None,
+            BUF_SIZE,
+            SpliceFFlags::empty(),
+        )?;
+        if res == 0 {
+            // We read 0 bytes from the input,
+            // which means we're done copying.
+            break;
         }
+        let _ = splice(pipe_rd, None, writer, None, BUF_SIZE, SpliceFFlags::empty())?;
```

Comment:
Again, not sure if this is necessary, but it might be better to keep track of how much has been read and written.  If not all of the data has been written at the end of the loop, you can just try to write the remainder at the end of the function until everything has been written.

Reply:
The code in the Rust stdlib falls back if the error occurs on the first `splice` usage. If an error occurs after that, the code panics.

I have done the same thing in https://github.com/uutils/coreutils/pull/1978/commits/e4d91f51e8e52e13a83ddb648d63af788d463c65, except instead of a panic we increment the error count and try to move on to the other files.